### PR TITLE
Headless importAll fails with error

### PR DESCRIPTION
**Pull request by umairsair**
importAll fails with error 'Project: .org.eclipse.egit.core.cmp already exists in the workspace!' and no further projects are imported.

The fix is to not go into special directories .metadata and .org.eclipse.egit.core.cmp. Moreover, importAll and removeAll do not fail immediately if operation on a project fails, instead it keeps on performing operation on all projects and at the end, headless builder returns with error code if operation of any project failed

### CODE REVIEW

**Review by jonahgraham — APPROVED** [review]

**Comment by jonahgraham** [comment]
Thanks @umairsair - LGTM. I'll try to remember to come back and merge this once CI completes. (I tried enabling [auto-merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request#about-auto-merge) but it doesn't seem to have done what I expect)

**Comment by umairsair** [comment]
BTW `importAll` using another workspace folder as argument will always fail with this error.. should we cater this `.org.eclipse.egit.core.cmp` project import failure in any way? but I think we should do the current fix anyways, all the projects should be tried to import and report all possible errors..

**Comment by jonahgraham** [comment]
> BTW importAll using another workspace folder as argument will always fail with this error.. should we cater this .org.eclipse.egit.core.cmp project import failure in any way? 

If you have time and energy to fix that case too, please go ahead and I will review/merge it.

> but I think we should do the current fix anyways, all the projects should be tried to import and report all possible errors..

+1

**Comment by umairsair** [comment]
> If you have time and energy to fix that case too, please go ahead and I will review/merge it.

any suggestions? I can think of ignoring .metadata folder.. to be on a safe side, we can additionally check that its a .metadata folder with version.ini or .log or .plugins folder inside it..

**Comment by jonahgraham** [comment]
Turns out I didn't understand the original problem. Can you just ignore `.metadata` and [`.org.eclipse.egit.core.cmp`](https://git.eclipse.org/c/egit/egit.git/tree/org.eclipse.egit.core/src/org/eclipse/egit/core/internal/efs/HiddenResources.java#n58) when recursing? i.e. treat those as reserved names that the importer can't import recursively.

**Comment by jonahgraham** [comment]
Do you want me to hold off submitting this PR until you look at this more?

**Comment by umairsair** [comment]
> Do you want me to hold off submitting this PR until you look at this more?

yes please! hold it

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java
Comment (jonahgraham) [inline build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java:374]: Can this check be moved into for loop a few lines down so it only applies to recursed directories?

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java
Comment (jonahgraham) [inline build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java:1048]: Use a set to make it a bit easier to add new entries later? (Untested code)

```suggestion
	private static Set<String> specials = Set.of(".metadata", ".org.eclipse.egit.core.cmp"); //$NON-NLS-1$ //$NON-NLS-2$
	private static boolean isSpecialDirectory(String directoryName) {
		return specials.contains(directoryName);
	}
```

Comment (jonahgraham) [reply build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java:1048]: You can probably then inline isSpecialDirectory too if you accept my suggested change.

File: build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java
Comment (umairsair) [reply build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java:374]: I was thinking the case of import as well where user can give regex matching `.metadata`.. should we not ignore in case of import?

Comment (jonahgraham) [reply build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/internal/core/HeadlessBuilder.java:376]: Sure - lets leave it where you have it.

**Review by jonahgraham — APPROVED** [review]

**Comment by jonahgraham** [comment]
Thanks @umairsair - LGTM and I hope this covers all your use cases. I'll try to remember to come back and merge this once CI completes - feel free to ping me once the build completes.

**Comment by umairsair** [comment]
@jonahgraham 

Can you please merge it

**Comment by jonahgraham** [comment]
Thanks @umairsair!

**Pull request merged**
